### DISCUSSION TITLE
Play/stop operator refactor

### DISCUSSION THREAD

**Discussion by tin2tin** [tin2tin/Audacity_Bridge_for_Blender#12] [Ideas]
A thread for discussion how to improve stop and playback keeping Blender and Audacity in sync.

Samy Tichadou did this patch, but we'll have to investigate more better ways to get the modal operator to work with/without start/stop buttons: https://github.com/tin2tin/audacity_tools_for_blender/pull/8

**Comment by samytichadou** [comment]
Ok so what would be the perfect behavior for this operator ?

To me, it would be some sort of toggle activating a "sync mode", where every play in blender would trigger a play in audacity, stop and pause also trigerring their audacity equivalent.

This could be achieved through a modal operator like this :
```
class ModalDrawOperator(bpy.types.Operator):
    """"""
    bl_idname = "view3d.modal_operator"
    bl_label = "Simple Modal View3D Operator"
    
    _playing = False
    _frame = 0
    _stop = False
    
    def execute(self, context):
        print("Playing = %s" % self._playing)
        print("Frame = %i" % self._frame)
        
        # play
        #if self._playing:
            # In Audacity :
            # set start frame
            # set end frame to blender range
            # play
            # set start frame to blender range
        # stop
        #elif not self._playing:
            #if self._stop:
                # In Audacity :
                # pause if posible
                # if not stop,
                # and find a way to set proper start frame and playhead frame
            #elif not self._stop:
                # In Audacity :
                # stop in audacity
                # set start frame

    def modal(self, context, event):
        # stop the operator
        if event.type == 'A': 
            # TODO replace this event with a property toggle to stop the modal from menu
            print("cancel")
            return {'CANCELLED'}
        
        # get a play command
        elif context.screen.is_animation_playing and self._playing == False:
            self._playing = True
            self._frame = context.scene.frame_current
            self.execute(context)
        
        # get a stop/pause command
        elif not context.screen.is_animation_playing and self._playing == True:
            self._playing = False
            # specify if stop or pause
            if self._frame == context.scene.frame_current:
                self._stop = True
            else:
                self._stop = False
            self._frame = context.scene.frame_current
            self.execute(context)
                
        return {'PASS_THROUGH'}

    def invoke(self, context, event):
        # set initial state (if playing in blender)
        if context.screen.is_animation_playing:
            self._playing = True
            self._frame = context.scene.frame_current
            self.execute(context)

        context.window_manager.modal_handler_add(self)
        return {'RUNNING_MODAL'}
```

The hard part would be to give the proper command to audacity to stay in sync with blender, even when stop/pause the play.
The playhead should be in the correct spot, but the range should also be correct, this is the tricky part (in the code, that is the commented part in the `execute` methode.

Another difficulty would be to warn the user when the sequencer or the strip is outdated and no more identical to audacity timeline content. This could be achieved by using some hash method to compare. This would return a serie of integer (always the same) corresponding to some properties of the strip/sequencer. 
For example, for a strip : 

```
def hash_sound_strip(strip):

    hash_settings = (
        strip.volume, #TODO check with fcurves
        strip.frame_final_duration,
        strip.frame_duration,
        
        strip.frame_offset_end,
        strip.frame_offset_start,

        strip.frame_still_end,
        strip.frame_still_start,
    )
    
    strip_hash = hash(hash_settings)
    
    return strip_hash
```
Here, we take some important properties to check if the length or the volume of the strip has changed, and return it as a hash. This hash could then be re process and compared a few moments later to be sure the strip didn't change. 
This hash would be processed when sending strip to audacity, and checked when launching play operator for example. If not identical, warning message would be triggered. It could also be re processed during the modal play, every n seconds, a check could happened, triggering warning if something changed and stopping the play operator.

Same could be done with sequence, but a little more complicated though, we'll have to test performances if we go with that.

Here are my test and retry for now, what do you think about all this ? (sorry for the length of this post :) )

**Comment by tin2tin** [comment]
You're right. As the play function is, it doesn't work in a good way at all. Stopping in Blender doesn't stop Audacity, if audacity already is at the end, so Audacity needs to play in a loop too, since in Blender it doesn't seem to be possible to switch loop-playing off. So Blender needs to define the range and tricker loop playing in Audacity, as you already stated. 

Or alternatively this snippet here could be used?
https://blender.stackexchange.com/questions/28869/how-to-disable-loop-playback-of-animation

As for mirroring playhead position, stip lengths and volume levels, the only way to get info from Audacity is the GetInfo command: https://manual.audacityteam.org/man/scripting_reference.html

And things like playhead position isn't supported. So we can't send the playhead position from audacity to blender. The default way of playing in Audacity is based on selection or a position. You select an area in the time code area, in a track area or a position in a track. 

(as a quick fix I'm adding loop playing right now, but the hack with changing the range(in sequence mode) after playing doesn't work, so playing in sequence mode will be off the 2. time around)

**Reply by samytichadou** [reply]
> So we can send the playhead position from audacity to blender

If the operator is a modal that could be achievable to track the audacity playhead, but this is some weird hack with a timer, not sure this is a good idea. We'll have to test it, but i think the accuracy of sync playback is not really achievable between audacity and blender, like with some ardour/jack pipeline or protools video-satellite workstation. I may be wrong (not a sound expert obviously) but i don't think audacity has enough hooks and way of communication for that...

**Reply by tin2tin** [reply]
I'm sorry it should have been: "So we **can't** send the playhead position from audacity to blender". https://forum.audacityteam.org/viewtopic.php?f=46&t=116136&start=20#p418454

Jack has been removed from recent Blender releases on Mac and Windows.

**Reply by tin2tin** [reply]
I wonder if a modal mode could detect the playhead being dragged in Blender  and stop the playback, and when it is released and starts playing again, a new play region is set-up and it is playing in sync again?

Small video demonstrating playback in sync:
https://blender.chat/file-upload/4rjJTpPHLK5M7NBXY/2021-03-16%2018-59-15.mp4

**Reply by samytichadou** [reply]
nice video ! yep this could be accessible through nodal, the idea would be to catch whenever blender is in a "playing state" after an event (mouse click, key pressed, even a timer (but mouse click and key press is enough in this situation)) and this would allow sending the appropriate command to audacity to set the region accordingly and start playing. 

It could be a delay, so we could also track frame changing when blender isn't playing and try to set them on the fly (timer could be an idea here) in audacity to be able to launch the play as soon as the key is pressed.
